Added Central Weather Administration (Taiwan) as a new source

chunshek opened the pull request:
I have added Central Weather Administration as a new national source for Taiwan. It works for both current location (reversed geocoded) and user-selected location (from searches).

papjul commented:
Does the location endpoint gives you whether the station supports UV index, normals and specific townships?

If that's the case, you could store a location parameter `supports_uv_index`, `supports_normals`, etc, instead of hardcoding the list of stations with their coordinates which is unmaintainable.

Otherwise, great job! I didn't test it, but it looks like you really understood how everything worked, even advanced features like location parameters.

chunshek commented:
> Does the location endpoint gives you whether the station supports UV index, normals and specific townships?
> 
> If that's the case, you could store a location parameter `supports_uv_index`, `supports_normals`, etc, instead of hardcoding the list of stations with their coordinates which is unmaintainable.

Unfortunately the location endpoint doesn't provide the current UV index observation or normal temperature ranges. I know that hard-coding these stations isn't "elegant", but that's the best I can do without dropping these data points entirely. Thankfully it's only a "small" number too, otherwise I agree with you that it can quickly become unmanageable.

According to the [API doc for the temperature normals endpoint](https://opendata.cwa.gov.tw/dataset/climate/C-B0027-001), that information will only be updated once every 10 years. So "maintaining the list" is hopefully not a huge issue.

As for the township warning areas - it's the same story: The location endpoint doesn't tell you. Among 368 townships in Taiwan, CWA classifies 82 as belonging to the special warning areas. Those are the ones in the `CWA_TOWNSHIP_WARNING_AREAS` list (not all 368). The townships are stable though - the last time there were changes to the number of townships in Taiwan was in 2004. If any are to be created or merged in the future, it's going to be huge news because it affects local and national politics and election cycles, and will be announced months ahead of time.

I think what I can offer is to comment these lines with human-readable names at the end of each line, so that they are easier for future maintenance if necessary. I'll push another commit and let you know.

> Otherwise, great job! I didn't test it (yet), but it looks like you really understood how everything worked, even advanced features like location parameters.

Thank you! It took a while to look under the hood. But now that I get how everything works, I think it's pretty amazing.

chunshek commented:
Hi @papjul! As discussed earlier, I have added comments to `CwaGeoLookup.kt`. In addition, I fixed an issue which caused parsing issues when combining other secondary sources for Alerts, AQI and Normals.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaGeoLookup.kt:
```diff
@@ -0,0 +1,217 @@
+// Information is updated once every 10 years. Last update was after 2020.
+//
+// Source (last checked 2024-05-29):
+// https://opendata.cwa.gov.tw/dataset/climate/C-B0027-001
```
Where do you see the information below on this link?

chunshek replied:
From the page in the link, the update frequency of this source is mentioned as such:
> 更新頻率 每10年

If you click on the 資料預覽 (Preview) or 資料範例 (Example) tabs on the same page, you'll see in the sample data that there are year ranges for the averages (in XML from "Example"):

```
<StationStartYear>1991</StationStartYear>
<StationEndYear>2020</StationEndYear>
```

Temperature normals should theoretically smooth out over long periods of time (10+ years) even for wider geographical ranges. But still these figures are recorded only for 20-something stations out of 700+. This is similar to the real-time UV Index issue below. So if we remove one, we should probably remove the other.

papjul replied:
By "below" I meant below my commented line. GitHub doesn't show it to you unfortunately depending on the page.

What I meant is where do you see the list with Banqiao, Tamsui, Anbu, etc, on this link? I only see Banqiao when clicking on one of the tabs.

chunshek replied:
Ah, no. I just pulled the entire set of data from the API endpoint, and extracted list of station IDs and coordinates from the output. No magic there. Pretty much brute force.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaGeoLookup.kt:
```diff
@@ -0,0 +1,217 @@
+// This list allows matching a location to the nearest of those 29 stations.
+//
+// Source (last checked 2024-05-29):
+// https://opendata.cwa.gov.tw/dataset/observation/O-A0003-001
```
This link shows that this goes further than just UVI, here is what is listed:
AirPressure、AirTemperature、GustInfo、WindDirection、PeakGustSpeed、WindSpeed、Precipitation、RelativeHumidity、SunshineDuration、UVI

Could you clarify why UV needs a special case? If some stations don't have all kind of data, we should not mix stations between each other as they can be opposite, and leave the missing information empty instead.

chunshek replied:
Unfortunately the CWA only record UV measurements at fewer than 30 stations out of the 700+. In theory the measurements could be valid over longer distances, as variations of solar radiation over longer distances should be less than those of other weather elements.

But I can see why you wouldn't feel comfortable including UV Index from nearby stations. I'm fine removing it. I -seem to remember Breezy would "estimate" the daytime UV Index from the forecast data – not ideal, but there are no ideal solutions here, only preferable ones. And I defer to your preference. 😉

BTW, the above is an older endpoint that does not take geographic coordinates, only station IDs, but it can output real-time UV observations. I did not use this endpoint for the main request because it would require downloading and looping through an entire 700+ station list to identify the nearest one, which you indicated in other threads that you do not want happening in the app.

The newer endpoint (used for the main request) takes geographic coordinates, but does not output UV Index measurements even if the station has it.

papjul replied:
> The newer endpoint (used for the main request) takes geographic coordinates, but does not output UV Index measurements even if the station has it.

I believe you should contact them to let them know that this information is missing, they probably already have the data and forgot to add it. That would be the easiest path for everyone

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaResultConverter.kt:
```diff
@@ -0,0 +1,645 @@
+    return WeatherWrapper(
+        current = getCurrent(weatherResult, uvResult, ignoreFeatures),
+        normals = if (!ignoreFeatures.contains(SecondaryWeatherSourceFeature.FEATURE_NORMALS)) getNormals(normalsResult) else null,
+        dailyForecast = getDailyForecast(weatherResult.data!!.aqi[0].town!!.daily, sunResult, moonResult),
```
Please safely throw an exception instead forcing Null Pointer Exceptions. You can see an example here: https://github.com/breezy-weather/breezy-weather/blob/v5.2.4/app/src/main/java/org/breezyweather/sources/openmeteo/OpenMeteoResultConverter.kt#L94-L97

chunshek replied:
Will do! Thanks for the pointer.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaResultConverter.kt:
```diff
@@ -0,0 +1,645 @@
+        "CO" -> 28.01
+        else -> return null
+    }
+    // assume 25C
```
I don't think you can assume it's always 25 °C in Taiwan

chunshek replied:
I will add temperature as a parameter to this function, so the concentration calculation can be more precise based on the reported ambient temperature.

I remember that the formula used assumes 1 atmospheric pressure as well. I'll see if I can find a more detailed formula that allows plugging in the air pressure. If so, I'll add that as a parameter also.

chunshek replied:
I figured out how one of the constants in the original formula was derived from the molar gas constant but applied at 1 atm (1013.25 hPa). I have update the function to take both temperature and air pressure into account.

I'm not going to make `convertSecondary` loop through the labyrinthine API response to get the temperature and air pressure. So if null values are passed, the temperature and air pressure will be assumed at 25°C and 1 atm (1013.25 hPa) respectively, which are [commonly used](https://en.wikipedia.org/wiki/Useful_conversions_and_formulas_for_air_dispersion_modeling) parameters.

papjul commented:
I added a few comments, I will let you check. It's not easy to review because there is a lot to review and the API is not really friendly, resulting in some dirty bits of code to workaround the way the API works (which is not your fault). Additionally, it's not documented in English, making my review really difficult.

chunshek commented:
Thanks for looking into it, Julien. I understand that there was a lot of code, the API outputs are very messy – I was envious when I saw how neat the other sources send their responses and how CWA's output is like a ball of yarn – and of course the language barrier didn't help either. (Désolé ! La langue chinoise est tellement difficile pour les locuteurs non natifs.)

I will first respond to each of the comments above, then patch my code in the next day or two, and let you know after I commit the patches.

chunshek commented on app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaGeoLookup.kt:
```diff
@@ -0,0 +1,217 @@
+//  • 蘭嶼綠島 Lanyu and Ludao ("L"): 2 townships
+//
+// Source of township specification (last checked 2024-25-29):
+// https://www.cwa.gov.tw/Data/js/info/Info_Warning.js
```
I pasted the wrong link! Sorry. It should be this one instead:
https://www.cwa.gov.tw/Data/js/info/Info_Town.js

I will correct it in my patch.

The township codes have been normalized from the 7-digit ones in the Javascript (old national standard) to 8-digit ones in the API response (current national standard).

chunshek commented:
To make sure we are both on the same page, here's a list of what I'll do:
- [ ] Remove current UV Index observation using the hardcoded list
- [ ] Remove temperature normals using the hardcoded list
- [ ] Correct the reference link for warning areas
- [ ] Add a check to throw `InvalidOrIncompleteDataException`
- [ ] Add temperature as parameter for more precise pollutant concentration conversion

papjul commented:
> Remove temperature normals using the hardcoded list

No, you can keep it. It's common to have normals on a different endpoint. It's better if an endpoint can give it to us rather than relying on a list to maintain, but if we do it like that, I must understand how to update that list, and that's what was unclear to me

chunshek commented:
Done patching! Over to you to review the new commits.

papjul commented on app/src/src_nonfreenet/org/breezyweather/sources/cwa/CwaResultConverter.kt:
```diff
@@ -73,11 +72,14 @@ fun convert(
     id: String,
     ignoreFeatures: List<SecondaryWeatherSourceFeature>
 ): WeatherWrapper {
+    if (weatherResult.data == null) {
```
Should be `weatherResult.data?.aqi?.getOrNull(0)?.town?.daily == null || weatherResult.data.aqi[0].town.hourly == null` (may need some `!!`), you're doing too many things after that can lead to Null Pointer Exceptions

chunshek replied:
Done.

papjul commented:
Alright, looks good to me, could you put together your 15 commits into 1 for easier integration and commit history?

If not familiar with Git, there are instructions here: https://github.com/breezy-weather/breezy-weather/blob/main/CONTRIBUTE.md#submit

Thanks!

chunshek commented:
Git is not my forté, so I created some oddities in the commit with my force push. I hope I haven't messed things up too much! In any case it's now just one commit instead of 15. Let me know if I need to change anything.

papjul commented on app/build.gradle.kts:
```diff
@@ -18,8 +18,8 @@ android {
-        versionCode = 50204
-        versionName = "5.2.4"
+        versionCode = 50205
+        versionName = "5.2.5"
```
Did you rebase your commit? This is not part of your changes.

papjul replied:
Your branch `main` is 82 commits behind current `main`

papjul replied:
```
git remote add upstream https://github.com/breezy-weather/breezy-weather
git fetch upstream
git rebase upstream main
```

papjul commented on app/src/main/java/org/breezyweather/common/source/LocationPreset.kt:
```diff
@@ -62,7 +62,8 @@ enum class LocationPreset(
     HONG_KONG("openmeteo", alert = "wmosevereweather", normals = "accu"),
-    ISRAEL("ims", airQuality = "openmeteo", pollen = "openmeteo", minutely = "openmeteo");
+    ISRAEL("ims", airQuality = "openmeteo", pollen = "openmeteo", minutely = "openmeteo"),
+    TAIWAN("cwa");
```
I will remove this after merging, as I cannot make users add a location and then find out they can't have fresh data because the source that is rate-limited; that's not a good experience.

I will still probably bundle an API key but it will only be usable at the beginning of the day until the daily limit is reached. Then, you have to wait for each daily reset. That's what is currently being done with OpenWeather.

chunshek replied:
Oh! I don't know how to bundle an API key without making it visible in the source code. I'll leave that with you.

The default allowance for CWA's API key is 20,000 calls and 2GB of transfer per day.

One can apply for an upgrade to advanced membership. The upgrade is free/gratis, but requires filling in name and phone number, as well as manual approval by CWA staff. The allowance for the advanced API key is 300,000 calls and 20GB of transfer per day.

papjul replied:
You can't. GitHub is adding the keys when it builds the release. We also prefer to only use keys registered by the project in official releases ;)

chunshek commented:
I had some issues with my local git history, and couldn't manage to use your instructions to rebase earlier. I dropped my older commits and I think I'm all synced up now…? Please retry.

papjul commented:
Merged

chunshek commented:
Thanks a million for all the trouble going through my code and guiding me through the process! I'll look into those comments later.
